File: 백준/Gold/14502. 연구소/연구소.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    static int N, M, answer;
    static int[][] graph;
    static List<Node> virusList = new ArrayList<>();
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    static int stoi(String s) {
        return Integer.parseInt(s);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = stoi(st.nextToken());
        M = stoi(st.nextToken());

        graph = new int[N][M];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                graph[i][j] = stoi(st.nextToken());

                if (graph[i][j] == 2) {
                    virusList.add(new Node(i, j));
                }
            }
        }

        dfs(0);

        System.out.println(answer);
    }

    static void dfs(int depth) {
        if (depth == 3) {
            int[][] tempGraph = new int[N][M];

            for (int i = 0; i < N; i++) {
                for (int j = 0; j < M; j++) {
                    tempGraph[i][j] = graph[i][j];
                }
            }

            for (Node virus : virusList) {
                bfs(virus, tempGraph);
            }

            int safeArea = 0;
            for (int[] row : tempGraph) {
                for (int x : row) {
                    if (x == 0) {
                        safeArea++;
                    }
                }
            }

            answer = Math.max(answer, safeArea);
            return;
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (graph[i][j] == 0) {
                    graph[i][j] = 1;
                    dfs(depth + 1);
                    graph[i][j] = 0;
                }
            }
        }
    }

    static void bfs(Node start, int[][] graph) {
        Queue<Node> q = new LinkedList<>();
        q.add(start);

        while (!q.isEmpty()) {
            Node now = q.poll();

            for (int i = 0; i < 4; i++) {
                int nx = now.x + dx[i];
                int ny = now.y + dy[i];

                if (!(0 <= nx && nx < N)) {
                    continue;
                }
                if (!(0 <= ny && ny < M)) {
                    continue;
                }
                if (graph[nx][ny] != 0) {
                    continue;
                }

                graph[nx][ny] = 2;
                q.add(new Node(nx, ny));
            }
        }
    }

    static class Node {

        int x, y;

        public Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}

File: 백준/Gold/14502. 연구소/연구소.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.ArrayDeque
import java.util.StringTokenizer

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine())
    
    val N = st.nextToken().toInt()
    val M = st.nextToken().toInt()

    val graph = Array(N) { IntArray(M) }
    val virusList = mutableListOf<Pair<Int, Int>>()

    repeat(N) { i ->
        st = StringTokenizer(br.readLine())

        repeat(M) { j ->
            graph[i][j] = st.nextToken().toInt()
            if (graph[i][j] == 2) {
                virusList += Pair(i, j)
            }
        }
    }

    val dx = intArrayOf(-1, 1, 0, 0)
    val dy = intArrayOf(0, 0, -1, 1)

    var answer = 0

    fun dfs(depth: Int) {
        if (depth == 3) {
            val temp = Array(N) { i -> graph[i].clone() }
            val q = ArrayDeque<Pair<Int, Int>>()
            virusList.forEach { q += it }

            while (q.isNotEmpty()) {
                val (x, y) = q.removeFirst()
                for (dir in 0..3) {
                    val nx = x + dx[dir]
                    val ny = y + dy[dir]
                    if (nx in 0 until N && ny in 0 until M && temp[nx][ny] == 0) {
                        temp[nx][ny] = 2
                        q += Pair(nx, ny)
                    }
                }
            }

            var safe = 0
            for (i in 0 until N) {
                for (j in 0 until M) {
                    if (temp[i][j] == 0) safe++
                }
            }
            answer = maxOf(answer, safe)
            return
        }

        for (i in 0 until N) {
            for (j in 0 until M) {
                if (graph[i][j] == 0) {
                    graph[i][j] = 1
                    dfs(depth + 1)
                    graph[i][j] = 0
                }
            }
        }
    }

    dfs(0)
    println(answer)
}
